### ISSUE TITLE
MapView doesnt work with async/await

### ISSUE THREAD

**Issue by GeorgeKnap** [Esri/jsapi-resources#76]
This probably applies to all interfaces that inherits from `corePromise`
If you try to return MapMiew instance in a promise using `async`& `await` you get compilation error:
`Return expression in async function does not have a valid callable 'then' member.`

**Comment by beginor** [comment]
Because `async/await` uses  ES6 's `Promise` interface , but ESRI jsapi uses dojo 's `IPromise` interface , they are not the same.

**Comment by GeorgeKnap** [comment]
Yeah it's not the same. I was wondering if devs from ESRI knows about this issue and whether or not they are going to support ES6.

**Comment by eric-parsons** [comment]
You can convert any promise-like object such as IPromise into an ES6 promise by using Promise.resolve:
```
await Promise.resolve(mapView);
```

However, there is another issue that you will immediately run into. Several objects in the JavaScript API are promises that fullfill with themselves when resolved. When attempting to resolve such promises, promise implementations that follow the [A+ promise specification](https://promisesaplus.com/) must recurse forever (or optionally detect the cycle and reject with a TypeError). This applies to most browser's native implementations and several popular polyfills such as Bluebird. If you run the above code in e.g. Google Chrome, it will simply hang forever. To get around this, I created the following helper function (written in TypeScript, which I infer is what you're using):

```
/**
 * Returns a promise that is fulfilled when the given Esri API object is ready.
 * @param object The object to wait for.
 */
export function ready(object: any): Promise<void> {
    return new Promise<void>((resolve, reject) => {
        if (object && typeof object.then === "function") {
            object.then(() => resolve(), reject);
        } else {
            resolve();
        }
    });
}
```

The trick is that it returns a `Promise<void>` instead of e.g. `Promise<MapView>`. The usage would then be:
```
await ready(mapView);
```

You might also find this helper useful as well:
```
/**
 * Loads the given Esri API object and returns a promise that is fulfilled when
 * the object is loaded.
 * @param object The object to load.
 */
export async function load(object: any): Promise<void> {
    if (object && typeof object.load === "function") {
        object = object.load();
    }

    await ready(object);
}
```

Then you can do this if you need to load e.g. a layer:
```
await load(layer);
```

**Comment by Ixonal** [comment]
Just ran into this issue, but with Layers, and could see this being a really annoying issue going forward. Would it be possible to alter the type definitions to specify that IPromise<T> inherits from TypeScript's Promise<T> or Thenable<T> interfaces?

**Comment by dasa** [comment]
Fixed in 4.6. See: https://blogs.esri.com/esri/arcgis/2017/12/14/making-better-promises/
